src/simple_lottery/init.rs

```diff
@@ -0,0 +1,94 @@
+//! Parameter derivation for simple lottery.
+
+use super::params::Params;
+use super::setup::Setup;
+
+use std::f64::consts::LN_2;
+
+/// Calculates Setup parameters.
+pub fn make_setup(params: &Params) -> Setup {
+    let ratio = params.set_size as f64 / params.lower_bound as f64;
+
+    // Execute binary search for the optimal value of `rs`.
+
+    let mut l: f64 = 0.0;
+    let mut r: f64 = ratio;
+
+    loop {
+        let m = (l + r) / 2.0;
+        let rs = m;
+        let rc = ratio / rs;
+
+        let lhs = params.soundness_param / (rs.ln() - 1.0 + 1.0 / rs);
+        let rhs = params.completeness_param / (rc - 1.0 - rc.ln());
+
+        if (m <= l) || (m >= r) {
+            let lhs = lhs * LN_2;
+            let rhs = rhs * LN_2;
+            let max = if lhs > rhs { lhs } else { rhs };
+            let u = max.ceil();
+            let mu = u * rc;
+            return Setup {
+                proof_size: u as u64,
+                lottery_probability: mu / params.set_size as f64,
+            };
+        }
+
+        if lhs > rhs {
+            l = m;
+        } else {
+            r = m;
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::super::params::Params;
+    use super::make_setup;
+
+    #[test]
+    fn basic() {
```

Comment:
You should merge these 3 tests in one, say `test_correctness`, and iter on an array of `(Setup, expected_u, expected_mu)`, as the test names ("basic" and "fractional ratio") do not give much information and, so far, you have not distinguished in the parameter generation the case where the lambdas are equal.

Reply:
+1
Also, we should adopt the style @rrtoledo used in [here](https://github.com/cardano-scaling/alba/blob/5c4b4c35c8565344fa4c01f6217c26ab11c5f9ed/src/simple_lottery.rs#L128)

Reply:
This is mostly a stylistic preference, but I prefer simple tests without loops since they give more confidence that they test the right thing (no bugs in the tests). Also, if a test fails, it is clear which one fails, unlike with a loop.

Reply:
There are no differences between the tests though (they test the exact same function, which has no forks/cases). It does not make sense to separate them. Especially if you want to add more values, it means writing another test instead of just adding a test vector.

Reply:
I changed the tests to use parameterized test instances which reduce code duplication. Other benefits:
1. no complicated logic in tests (loops)
2. each test case is treated as its own test, so if a test fails, you know which one
```running 5 tests
test simple_lottery::init::tests::all::basic ... ok
test simple_lottery::init::tests::all::different_lambdas ... ok
test simple_lottery::init::tests::all::small_lambda_reliability ... ok
test simple_lottery::init::tests::all::fractional_ratio ... ok
test simple_lottery::init::tests::all::small_lambda_security ... ok
```
Take a look!

Reply:
Adding more tests now is indeed easier, thanks a lot for that.

This change however does not change the fact that you are not testing different cases of the code, or checking explicitely that the lambdas are computed correctly. Hence the way you split the tests seem artificial (why checking in particular smalll lambda reliability if the formula is the same regardless on how high or low this lambda is?).

If you do want to check something precisely, I would encourage you to use loops contrary to what you say. Having only one test_vector is putting the code at risk of a false positive, the code could also be tailored for this specific value.

Reply:
> This change however does not change the fact that you are not testing different cases of the code

Can you elaborate what you mean? What cases do I not check?

> checking explicitely that the lambdas are computed correctly

lambdas are not computed, they are inputs, I'm not sure what you mean

> why checking in particular smalll lambda reliability

to test that the binary search moves in the right direction

> If you do want to check something precisely, I would encourage you to use loops

Could you explain what this loop would look like? What does iterate over and what does it check in each iteration? Pseudo-code would be ideal.

> the code could also be tailored for this specific value

This can be true of any code. Tests generally do not aim to test all possible inputs. The goal is only to test different "behaviors" of the code (including all edge cases).

Reply:
I like randomized tests but they never replace basic deterministic unit tests. In this case, what would you check in randomized tests? I still don't see what you are proposing exactly.